numpy/_core/src/multiarray/stringdtype/dtype.c

```diff
@@ -185,6 +185,41 @@ _eq_comparison(int scoerce, int ocoerce, PyObject *sna, PyObject *ona)
     return na_eq_cmp(sna, ona);
 }
 
+// currently this can only return 1 or -1, the latter indicating that the
+// error indicator is set
+NPY_NO_EXPORT int
+stringdtype_compatible_na(PyObject *na1, PyObject *na2) {
```

Comment:
This is only used by `string_comparison_resolve_descriptors`, in principle I could expose `na_eq_cmp` but exposing this seemed more useful.

Reply:
Agreed this is more useful. My sense, though, would be to make this function like your next one, i.e., have its signature include an `PyObject **out_na`, and, if this is non-NULL, return the compatible na in it. So, something like,
```
NPY_NO_EXPORT int
stringdtype_compatible_na(PyObject *na1, PyObject *na2, PyObject **out_na)
{
    if (na1 != NULL && na2 != NULL) {
        int na_eq = na_eq_comp(na1, na2);
        if (na_eq < 0) {
            return -1;
        }
        else if (na_eq == 0) {
            PyErr_Format(PyExc_TypeError,
                     "Cannot find a compatible null string value for "
                     "null strings '%R' and '%R'", na1, na2);
            return -1;
        }
    }
    if (out_na != NULL) {
        *out_na = na1 ? na1 : na2;
    }
    return 0;
}
```
I think with that, you don't really need the next function any more - in the few places where it is used you can just calculate `coerce` directly.

Reply:
Applied this suggestion in the most recent commit.